### PR TITLE
FormatWriter: implement granular blanks control

### LINKED ISSUE #2511: Feature request: ability to have multiple blank lines around various kinds of elements
- **Version**: 3.0.0-RC2

## Desire

I should like to be able to configure Scalafmt to format code to look like this, with two blank lines surrounding top-level [template definitions (`TmplDef)`](https://dotty.epfl.ch/docs/reference/syntax.html#declarations-and-definitions):

```scala
import scala.collection.mutable.ArrayBuffer


/** 2 blank lines above and below top-level trait definition */
trait A


// 2 blank lines above top-level class definition
// (none below due to end of file)
class B:

  /** 1 blank line above and below method definition */
  def A = ???

  /** 1 blank line above and below nested trait definition */
  trait Bar
      extends A

  /** 1 blank line above (none below due to end of file) */
  lazy val x = 1
```

Some coding styles may prefer something like this, with three blank lines surrounding top-level template definitions and two blank lines surrounding nested method definitions:

```scala
import scala.collection.mutable.ArrayBuffer



/** 3 blank lines above and below top-level trait definition */
trait A:

   // 1 blank line above (2 below due to following def)
   val x = 32
  
  
   /** 2 blank lines above and below nested method definition */
   def A =
      // beginning of a lengthy method body containing a blank line
     
      // continuation of lengthy body following a blank line
      ???


   // 2 blank lines above nested method definition
   // (3 below due following class)
   def B = ???



// 3 blank lines above top-level class definition
// (none below, due to end of file)
class B:


   /** 2 blank lines above and below */
   def A = ???


   /** 2 blank lines above and below nested-trait */
   trait Bar
      extends A


   // 1 blank line normally above val, but 2 enforced due to
   // preceding nested trait (none following due to end of file)
   lazy val x = 1
```

## Rationale

Being able to add more than one blank line before certain major elements such as [template definitions (`TmplDef)`](https://dotty.epfl.ch/docs/reference/syntax.html#declarations-and-definitions) would make it easier to distinguish between the various major and minor elements in a source file. For example, it is a lot easier to see where a class begins and ends if that class is surrounded by more whitespace than the methods and other items _within_ that class. In Scala, where multiple traits, classes, objects, etc. are often defined in a single source file, marking the boundaries of such major elements clearly would be especially useful.

Similarly, it is not unusual in many Scala coding styles to allow single blank lines within a method body. It would therefore be useful to be able to configure, say, two blank lines before and after method definitions, so that the boundaries of the method are clear and not obfuscated by the the intra-method blank lines.

## Notes

1. This feature would be useful even if it addressed only top-level template definitions.

2. However, it would be beneficial to be able to configure the number of blank lines surrounding many of the different Scala 3 syntax [Declarations and Definitions](https://dotty.epfl.ch/docs/reference/syntax.html#declarations-and-definitions) (e.g. `RefineDcl`, `TmplDef`, `Extension`).

3. It may be useful to be able to configure the minimum number of blank lines surrounding top-level elements of each kind separately from those same elements when they are nested (e.g. within the body of a template or method).

4. It may also be useful to be able to treat single-line declarations and definitions separately from those that extend over several lines.

### PR DESCRIPTION
**Pull request by kitbellew**
Fixes #2511.

### CODE REVIEW

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala [hunk 1]
```diff
@@ -1085,33 +1087,55 @@ class FormatWriter(formatOps: FormatOps) {
...
+      @tailrec
+      def getNest(tree: Tree, curNest: Int): Int = tree.parent match {
+        case Some(_: Source) | None => curNest
+        case Some(t: Pkg) if !indentedPackage(t) => getNest(t, curNest)
+        case Some(t) => getNest(t, curNest + 1)
+      }
+      def setTopStats(owner: Tree, stats: Seq[Stat], curNest: Int = 1): Unit = {
+        if (stats.isEmpty) return
+        val nest = getNest(owner, curNest)
+        if (nest < 0) return
```
Comment (tgodzik) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:1109]: When will nest be smaller than 0 ? It seems it's only incremented.

**Review by tgodzik — APPROVED** [review]
Just one small question, otherwise LGTM!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala (hunk 1 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:1109]: I used to have a check for `Term.Block` there, which returned -1, but removed it as unnecessary (see comment below). however, in the future, it's possible that we'll add another check which will prohibit nesting by returning a negative value.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala [hunk 2]
```diff
@@ -1134,6 +1158,9 @@ class FormatWriter(formatOps: FormatOps) {
         override def apply(tree: Tree): Unit =
           tree match {
             case _: Term.Block =>
+            case t: Source =>
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:1161]: the Block is skipped just above (in connection with your getNest question).

**Pull request merged**